ding/policy/diayn.py

```diff
@@ -0,0 +1,624 @@
+            logits = self._learn_model.forward(next_obs, mode='compute_discriminator')['q_discriminator']
+        else:
+        '''
+        for _ in range(self.cfg.discriminator_over_learning):
+            loss_discriminator = torch.nn.CrossEntropyLoss(reduction='mean')
+            logits = self._learn_model.forward(next_obs_without_skills, mode='compute_discriminator')['q_discriminator']
+            discriminator_loss = loss_discriminator(
+                logits,
+                torch.nonzero(sklls_batch)[:, 1]
+            )  # z_one_hot_reduce is z_one_hot without one hot (N, C) --> (N)
+            loss_dict['discriminator_loss'] = discriminator_loss
+
+            # update discriminator network
+            self._optimizer_discriminator.zero_grad()
+            loss_dict['discriminator_loss'].backward()
+            self._optimizer_discriminator.step()
+
+        # =============
+        # after update
+        # =============
+        self._forward_learn_cnt += 1
+        # target update
+        self._target_model.update(self._learn_model.state_dict())
+        return {
+            'cur_lr_q': self._optimizer_q.defaults['lr'],
+            'cur_lr_p': self._optimizer_policy.defaults['lr'],
+            'priority': td_error_per_sample.abs().tolist(),
+            'td_error': td_error_per_sample.detach().mean().item(),
+            'alpha': self._alpha.item(),
+            'target_value': target_value.detach().mean().item(),
+            'reward_revised_mean': reward_revised.detach().mean().item(),
+            'discriminator_mean': discriminator_for_plotting.detach().mean().item(), 
+            'discriminator_loss': discriminator_loss.detach().item(),
+            'which_skills': skill_index_for_plotting,
+            **loss_dict
+        }
+
+    def _state_dict_learn(self) -> Dict[str, Any]:
+        ret = {
+            'model': self._learn_model.state_dict(),
+            'target_model': self._target_model.state_dict(),
+            'optimizer_q': self._optimizer_q.state_dict(),
+            'optimizer_policy': self._optimizer_policy.state_dict(),
+        }
+        if self._value_network:
+            ret.update({'optimizer_value': self._optimizer_value.state_dict()})
+        if self._auto_alpha:
+            ret.update({'optimizer_alpha': self._alpha_optim.state_dict()})
+        return ret
+
+    def _load_state_dict_learn(self, state_dict: Dict[str, Any]) -> None:
+        self._learn_model.load_state_dict(state_dict['model'])
+        self._target_model.load_state_dict(state_dict['target_model'])
+        self._optimizer_q.load_state_dict(state_dict['optimizer_q'])
+        if self._value_network:
+            self._optimizer_value.load_state_dict(state_dict['optimizer_value'])
+        self._optimizer_policy.load_state_dict(state_dict['optimizer_policy'])
+        if self._auto_alpha:
+            self._alpha_optim.load_state_dict(state_dict['optimizer_alpha'])
+
+    def _init_collect(self) -> None:
```

Comment:
some methods like `_init_collect` can reuse them in `SACPolicy`

Reply:
Thank you very much, I will delete parts that can be inherited from SAC